dataplane/src/nat/fabric.rs

```diff
@@ -0,0 +1,419 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+use crate::nat::prefixtrie::{PrefixTrie, TrieError};
+
+use net::vxlan::Vni;
+use routing::prefix::Prefix;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::net::IpAddr;
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+pub struct Pif {
+    name: String,
+    endpoints: Vec<Prefix>,
+    ips: Vec<Prefix>,
+    vpc: String,
+    peerings: Vec<String>,
+}
+
+impl Pif {
+    #[tracing::instrument(level = "trace")]
+    pub fn new(name: String, vpc: String) -> Self {
+        Self {
+            name,
+            endpoints: Vec::new(),
+            ips: Vec::new(),
+            vpc,
+            peerings: Vec::new(),
+        }
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn vpc(&self) -> &String {
+        &self.vpc
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn iter_endpoints(&self) -> impl Iterator<Item = &Prefix> {
+        self.endpoints.iter()
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn iter_ips(&self) -> impl Iterator<Item = &Prefix> {
+        self.ips.iter()
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn iter_peerings(&self) -> impl Iterator<Item = &String> {
+        self.peerings.iter()
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn add_endpoint(&mut self, endpoint: Prefix) {
+        self.endpoints.push(endpoint);
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn add_ip(&mut self, ip: Prefix) {
+        self.ips.push(ip);
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn add_peering(&mut self, peering: String) {
+        self.peerings.push(peering);
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn find_prefix(&self, ip: &IpAddr) -> Option<&Prefix> {
+        self.iter_endpoints().find(|&prefix| prefix.covers_addr(ip))
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct PeeringPolicy {
+    name: String,
+    vnis: [Vni; 2],
```

Comment:
Shouldn't this be a pair of VPCs, instead of their VNIs?

Reply:
Here the VNIs are keys to retrieve the VPCs, so indirectly, that's what we have, sort of. I suppose it would be more explicit to have VPC names, although I don't like referencing them by names a lot.

Ideally, I'd like to have a direct reference to the `Vpc` object here, but this requires some wrapping with refcounts/Cells and so on, as the VPC objects are already owned by the `GlobalContext` object. Same things for the PIFs, I'd like to get a direct reference to them from the `PeeringPolicy` and save the hashmap lookup. I expect to do more work in that direction, once we agree the overall structure is acceptable and this PR gets merged.